sgm/inference/helpers.py

```diff
@@ -276,30 +314,74 @@ def do_img2img(
                 if skip_encode:
                     z = img
                 else:
-                    z = model.encode_first_stage(img)
+                    with SwapToDevice(model.first_stage_model, device):
+                        z = model.encode_first_stage(img)
+
                 noise = torch.randn_like(z)
+
                 sigmas = sampler.discretization(sampler.num_steps)
                 sigma = sigmas[0].to(z.device)
 
                 if offset_noise_level > 0.0:
                     noise = noise + offset_noise_level * append_dims(
                         torch.randn(z.shape[0], device=z.device), z.ndim
                     )
-                noised_z = z + noise * append_dims(sigma, z.ndim)
-                noised_z = noised_z / torch.sqrt(
-                    1.0 + sigmas[0] ** 2.0
-                )  # Note: hardcoded to DDPM-like scaling. need to generalize later.
+                if add_noise:
+                    noised_z = z + noise * append_dims(sigma, z.ndim).cuda()
+                    noised_z = noised_z / torch.sqrt(
+                        1.0 + sigmas[0] ** 2.0
+                    )  # Note: hardcoded to DDPM-like scaling. need to generalize later.
+                else:
+                    noised_z = z / torch.sqrt(1.0 + sigmas[0] ** 2.0)
 
                 def denoiser(x, sigma, c):
                     return model.denoiser(model.model, x, sigma, c)
 
-                samples_z = sampler(denoiser, noised_z, cond=c, uc=uc)
-                samples_x = model.decode_first_stage(samples_z)
-                samples = torch.clamp((samples_x + 1.0) / 2.0, min=0.0, max=1.0)
+                with SwapToDevice(model.denoiser, device):
+                    with SwapToDevice(model.model, device):
+                        samples_z = sampler(denoiser, noised_z, cond=c, uc=uc)
+
+                with SwapToDevice(model.first_stage_model, device):
+                    samples_x = model.decode_first_stage(samples_z)
+                    samples = torch.clamp((samples_x + 1.0) / 2.0, min=0.0, max=1.0)
 
                 if filter is not None:
                     samples = filter(samples)
 
                 if return_latents:
                     return samples, samples_z
                 return samples
+
+
+@contextlib.contextmanager
+def SwapToDevice(
```

Comment:
Shouldn't this be `swap_to_device` to follow Python standard for naming functions?

This is replacing the style from streamlit_helpers.py?
```
                load_model(model.denoiser)
                load_model(model.model)
                samples_z = sampler(denoiser, randn, cond=c, uc=uc)
                unload_model(model.model)
                unload_model(model.denoiser)
```

Rather than checking a `lowvram_mode` it assumes the model will start not on the requested device so then move it there and back. It does have to do quite a bit more checks than the manual load/unload version and the torch.cuda.empty_cache() side effect might be aggressive here so I have mixed feelings but if it's what's needed to fit on smallest VRAM possible makes sense. 

I didn't spot it but a lot of VRAM can be saved by sharing VAE and ViT-bigG CLIP embedder between base and refiner if that hasn't also been included here yet.

Reply:
The idea is to "just work" if the models have or haven't been moved to the right device already, and put them back where they started. The cache emptying was copied from the original code and only triggers if the move actually happened. There are probably better ways to save memory usage (as you mention) so I wanted to avoid formalizing this too much yet, but I've gone ahead and added a param to `SamplingPipeline` to make it a little more explicit and obvious.

Reply:
The name should be `swap_to_device`, agreed.